### PR TITLE
rewrite, remove unused or broken stuff, copy firmware when copying kmod, add more panic arguments

### PR DESCRIPTION
**Pull request by YukariChiba**

### CODE REVIEW

File: doc/tinyramfs.5.scd [hunk 1]
```diff
@@ -22,102 +22,107 @@ Options which expect a device as a value can also refer to the device via
...
-	The utility to use for compressing the initramfs. The option accepts
-	arbitrary commands (eg. 'gzip -9'). Use 'cat' for an uncompressed
-	initramfs.
+root
 
-hostonly
+	The device your root filesystem is located on.
 
-	(bool) Whether or not to only copy modules specific to your system.
+	(available format: /dev/sdX, LABEL=label, PARTUUID=partuuid, UUID=uuid)
```
Comment (ziyao233) [inline doc/tinyramfs.5.scd]: Add a note for the "dangerous" `/dev/sdX`?

File: doc/tinyramfs.8.scd [hunk 2]
```diff
@@ -57,4 +57,4 @@ unlocking encrypted drives, mounting needed filesystems, etc.
 # AUTHORS
 
 tinyramfs by illiliti *https://github.com/illiliti*++
-man page by fluorescent_haze *https://github.com/fluorescent-haze*
+man page by fluorescent_haze *https://github.com/fluorescent-haze*
```
Comment (ziyao233) [inline doc/tinyramfs.8.scd]: What changes here? A missing newline?

File: hook/live/live.init [hunk 3]
```diff
@@ -14,40 +15,42 @@ mkdir -p "$medium" "$system" /run/initramfs/overlayfs/work /run/initramfs/overla
 
 live_disk_fs=${live_disk_fs:-iso9660}
 
-mount -t ${live_disk_fs} -o ro "/dev/disk/by-$live_disk_type/$live_disk" "$medium" || panic
+mount -t "${live_disk_fs}" -o ro "/dev/disk/by-$live_disk_type/$live_disk" "$medium" || panic "unable to mount live medium"
```
Comment (ziyao233) [inline hook/live/live.init]:
```suggestion
mount -t "$live_disk_fs" -o ro "/dev/disk/by-$live_disk_type/$live_disk" "$medium" || panic "unable to mount live medium"
```

File: hook/plymouth/plymouth [hunk 4]
```diff
@@ -1,12 +1,29 @@
...
-if [ -z $plymouth_nocopymodule ] && [ -z $monolith ] && [ -z $hostonly ]; then
-  cd "${moddir}/${kernel}/kernel" && \
-    find drivers/gpu/drm -type f 2> /dev/null |
+if [ -f "/usr/lib/plymouth/plymouth-populate-initrd" ]; then
+
+  /usr/lib/plymouth/plymouth-populate-initrd -t "${tmpdir}"
+
+  if [ -z "$plymouth_nocopymodule" ] && [ "$modules_copy" != "off" ]; then
+    if [ "$modules_copy" = "all" ]; then
+      cd "${moddir}/${kernel}/kernel" && \
```
Comment (ziyao233) [inline hook/plymouth/plymouth]:
```suggestion
      cd "$moddir/$kernel/kernel" && \
```

File: hook/plymouth/plymouth [hunk 5]
```diff
@@ -1,12 +1,29 @@
...
+          esac 2> /dev/null
       done
-fi
-
-if [[ -x /usr/lib/plymouth/plymouth-populate-initrd ]]; then
-  /usr/lib/plymouth/plymouth-populate-initrd -t ${tmpdir}
+    fi
+  fi
+else
+  print "plymouth is not installed, ignore hook"
```
Comment (ziyao233) [inline hook/plymouth/plymouth]: ignore -> skip

File: lib/common.sh [hunk 6]
```diff
@@ -93,17 +92,49 @@ copy_exec()
...
     done
 }
 
 copy_kmod()
 {
-    modprobe -S "$kernel" -D "$1" 2> /dev/null |
+    # prevent modules to be copied even in hooks
+    if [ "$modules_copy" = "off" ]; then
+        return
+    fi
```
Comment (ziyao233) [inline lib/common.sh]: `[ "$modules_copy"  = "off" ] && return`

File: lib/init.sh [hunk 7]
```diff
@@ -44,23 +44,32 @@ parse_cmdline()
...
-	--)           init_args+=${cmdline#*--}; break ;;
-	quiet)        init_args+=" --quiet" ;;
+        --)           init_args="${init_args}${cmdline#*--}"; break ;;
+        quiet)        init_args="${init_args} --quiet" ;;
 
+        # internal functions
         rdpanic) trap - EXIT ;;
         rddebug) set -x ;;
-	break) break_init=1 ;;
+        break) break_init=1 ;;
```
Comment (ziyao233) [inline lib/init.sh]: `debug` may be a more common name. Rename `rddebug` to `verbose`, and `break` to `debug`?

We should also document these options (could in another PR)

File: tinyramfs [hunk 8]
```diff
@@ -87,12 +87,7 @@ copy_hooks()
 
 copy_modules()
 {
-    # Skip this function if kernel
-    # compiled with builtin modules.
-    if [ "$monolith" ]; then
-        return
-
-    elif [ "$hostonly" ]; then
+    if [ "$modules_copy" = "hostonly" ]; then
```
Comment (ziyao233) [inline tinyramfs]: make these ifs a case?

**Review by ziyao233 — CHANGES_REQUESTED** [review]
Mostly okay, for the new `modules` option, we should provide a smooth way for users to migrate

File: doc/tinyramfs.5.scd [hunk 9]
```diff
@@ -22,102 +22,110 @@ Options which expect a device as a value can also refer to the device via
...
-keymap_path
+	Arguments passed to mount command when mounting root device.
 
-	Path to your keymap.
+rorw
 
-## LUKS
+	Define root partition ro/rw mounting, default to use ro to make use of init system.
+	
+	(available options: ro, rw)
```
Comment (ziyao233) [inline doc/tinyramfs.5.scd:53]: rename to `root_readonly` and use it as a boolean value?

**Review by ziyao233 — APPROVED** [review]

File: doc/tinyramfs.5.scd (hunk 9 above)
Comment (YukariChiba) [reply doc/tinyramfs.5.scd:53]: I would suggest keeping these two params to provide compatibility. A lot of users take well-known parameter for granted.

**Pull request merged**